pkg/state/appender.go

```diff
@@ -612,6 +612,108 @@ func (a *txAppender) appendTx(tx proto.Transaction, params *appendTxParams) (txS
+) (*proto.BlockSnapshot, crypto.Digest, error) {
+	blockInfo, errBlockInfo := a.currentBlockInfo()
+	if errBlockInfo != nil {
+		return nil, crypto.Digest{}, errBlockInfo
+	}
+	for i, txs := range params.snapshot.TxSnapshots {
+		txID, idErr := params.transactions[i].GetID(a.settings.AddressSchemeCharacter)
+		if idErr != nil {
+			return nil, crypto.Digest{}, idErr
+		}
+		if len(txs) == 0 { // sanity check
+			return nil, crypto.Digest{}, errors.Errorf("snapshot of txID %q cannot be empty", base58.Encode(txID))
+		}
+		txSh, shErr := calculateTxSnapshotStateHash(hasher, txID, blockInfo.Height, stateHash, txs)
+		if shErr != nil {
+			return nil, crypto.Digest{}, errors.Wrapf(shErr, "failed to calculate tx snapshot hash for txID %q at height %d",
+				base58.Encode(txID), blockInfo.Height,
+			)
+		}
+		stateHash = txSh
+		regSnapshots := txSnapshot{regular: txs}
+		if err := regSnapshots.Apply(a.txHandler.sa, params.transactions[i], false); err != nil {
+			return nil, crypto.Digest{}, errors.Wrap(err, "failed to apply tx snapshot")
+		}
+	}
+	return params.snapshot, stateHash, nil
+}
+
+func (a *txAppender) appendTxs(
+	params *appendBlockParams,
+	info *checkerInfo,
+	stateHash crypto.Digest,
+	hasher *txSnapshotHasher,
+) (*proto.BlockSnapshot, crypto.Digest, error) {
+	blockInfo, errBlockInfo := a.currentBlockInfo()
+	if errBlockInfo != nil {
+		return nil, crypto.Digest{}, errBlockInfo
+	}
+	blockV5Activated, err := a.stor.features.newestIsActivated(int16(settings.BlockV5))
+	if err != nil {
+		return nil, crypto.Digest{}, err
+	}
+	consensusImprovementsActivated, err := a.stor.features.newestIsActivated(int16(settings.ConsensusImprovements))
+	if err != nil {
+		return nil, crypto.Digest{}, err
+	}
+	blockRewardDistributionActivated, err := a.stor.features.newestIsActivated(int16(settings.BlockRewardDistribution))
+	if err != nil {
+		return nil, crypto.Digest{}, err
+	}
+	invokeExpressionActivated, err := a.stor.features.newestIsActivated(int16(settings.InvokeExpression))
+	if err != nil {
+		return nil, crypto.Digest{}, err
+	}
+	// Check and append transactions.
+	var bs proto.BlockSnapshot
+
+	for _, tx := range params.transactions {
+		appendTxArgs := &appendTxParams{
+			chans:                            params.chans,
+			checkerInfo:                      info,
+			blockInfo:                        blockInfo,
+			block:                            params.block,
+			acceptFailed:                     blockV5Activated,
+			blockV5Activated:                 blockV5Activated,
+			rideV5Activated:                  info.rideV5Activated,
+			rideV6Activated:                  info.rideV6Activated,
+			consensusImprovementsActivated:   consensusImprovementsActivated,
+			blockRewardDistributionActivated: blockRewardDistributionActivated,
+			invokeExpressionActivated:        invokeExpressionActivated,
+			validatingUtx:                    false,
+			currentMinerPK:                   params.block.GeneratorPublicKey,
+		}
```

Comment:
Is it necessary to use new `appendTxParams` for each transaction? I think it can be created before `for` cycle and used for all transactions in the block.